runtime/src/bank.rs

```diff
@@ -1737,19 +1737,27 @@ impl Bank {
             }
         }
 
-        if !already_genuine_program_exists {
+        if must_replace {
+            assert!(
+                already_genuine_program_exists,
+                "There is no program account to replace"
+            );
+        } else if !already_genuine_program_exists {
             assert!(
                 !self.is_frozen(),
                 "Can't change frozen bank by adding not-existing new native program ({}, {}). \
                 Maybe, inconsistent program activation is detected on snapshot restore?",
                 name,
                 program_id
             );
```

Comment:
Could you move this assertion to reside immediately before `store_account`? This assertion is tightly associated to `store_account`.  We can safely `assert!(!self.is_frozen())` even if `must_replace` is `true`. In general, we must not `store_account` once frozen in every conceivable situation. But that kind of protection isn't easy to add inside `store_account` itself, unfortunately. So, we're manually doing this because these activations deal with primitive arrangements with potentinal misbehavior of wracking frozen (snapshot) banks incorrectly.

So, the change hunk could be changed to look like this:

```rust
if must_replace {
  assert!(already_genuine_program_exists, ...)
} else if already_genuine_program_exists {
  return;
}

assert!(!self.is_frozen(), ...);
self.store_account(...);

```

Reply:
Moving this assert breaks a bunch of snapshot tests once a `NewVersion` builtin is added. What do you recommend?

Reply:
@CriesofCarrots Blame me, I completely forget that this fn must retain idempotent of AccountsDB; so more subtle adjustment was needed: https://github.com/solana-labs/solana/pull/13403/commits/2a9ff42055427c632d4eb2d81ea26b609a14f432 Hope this commit explains well. If still not sure, please say so. I'll update the in-source comment.

Sorry for confusing you because of my unreadable code. ;)